zebra-network/src/meta_addr.rs

```diff
@@ -15,17 +15,112 @@ use zebra_chain::serialization::{
+/// [`AddressBook::maybe_connected_peers`] and
+/// [`AddressBook::reconnection_peers`].
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum PeerConnectionState {
+    /// The peer has sent us a valid message.
+    ///
+    /// Peers remain in this state, even if they stop responding to requests.
+    /// (Peer liveness is derived from the `last_seen` timestamp, and the current
+    /// time.)
+    Responded,
+
+    /// The peer's address has just been fetched from a DNS seeder, or via peer
+    /// gossip, but we haven't attempted to connect to it yet.
+    NeverAttempted,
+
+    /// The peer's TCP connection failed, or the peer sent us an unexpected
+    /// Zcash protocol message, so we failed the connection.
+    Failed,
+
+    /// We just started a connection attempt to this peer.
+    AttemptPending,
+}
+
+impl Default for PeerConnectionState {
+    fn default() -> Self {
+        PeerConnectionState::NeverAttempted
+    }
+}
+
+impl Ord for PeerConnectionState {
+    /// `PeerConnectionState`s are sorted in approximate reconnection attempt
+    /// order, ignoring liveness.
+    ///
+    /// See [`CandidateSet`] and [`MetaAddr::cmp`] for more details.
+    fn cmp(&self, other: &Self) -> Ordering {
+        use PeerConnectionState::*;
+        match (self, other) {
+            (Responded, Responded)
+            | (NeverAttempted, NeverAttempted)
+            | (Failed, Failed)
+            | (AttemptPending, AttemptPending) => Ordering::Equal,
+            // We reconnect to `Responded` peers that have stopped sending messages,
+            // then `NeverAttempted` peers, then `Failed` peers
+            (Responded, _) => Ordering::Less,
+            (_, Responded) => Ordering::Greater,
+            (NeverAttempted, _) => Ordering::Less,
+            (_, NeverAttempted) => Ordering::Greater,
+            (Failed, _) => Ordering::Less,
+            (_, Failed) => Ordering::Greater,
+            // AttemptPending is covered by the other cases
+        }
+    }
+}
+
+impl PartialOrd for PeerConnectionState {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
 /// An address with metadata on its advertised services and last-seen time.
 ///
 /// [Bitcoin reference](https://en.bitcoin.it/wiki/Protocol_documentation#Network_address)
 #[derive(Copy, Clone, Debug, Eq, PartialEq)]
 pub struct MetaAddr {
     /// The peer's address.
     pub addr: SocketAddr,
+
     /// The services advertised by the peer.
+    ///
+    /// The exact meaning depends on `last_connection_state`:
+    ///   - `Responded`: the services advertised by this peer, the last time we
+    ///      performed a handshake with it
+    ///   - `NeverAttempted`: the unverified services provided by the remote peer
+    ///     that sent us this address
+    ///   - `Failed` or `AttemptPending`: unverified services via another peer,
+    ///      or services advertised in a previous handshake
+    ///
+    /// ## Security
```

Comment:
Nice

Reply:
Yeah we should do a follow-up where we make the verified and unverified timestamps and services into different fields.

Reply:
This is one of those distributed systems things - distinguishing untrusted data (from peers) from verified data (that we've created or checked outselves).

Reply:
I have a bunch of these kinds of changes, let's talk about them later this week.